de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java

```diff
@@ -0,0 +1,273 @@
+    private final ISharedEditorListener listener = new AbstractSharedEditorListener() {
+        @Override
+        public void editorActivated(User user, SPath file) {
+            fileOpened(file);
+        }
+    };
+
+    private List<StartHandle> stoppedUsers = null;
+    private User remoteUser = null;
+
+    public InstantOutgoingProjectNegotiation(final JID peer, //
+        final List<IProject> projects, //
+        final ISarosSessionManager sessionManager, //
+        final ISarosSession session, //
+        final IEditorManager editorManager, //
+        final IWorkspace workspace, //
+        final IChecksumCache checksumCache, //
+        final XMPPConnectionService connectionService, //
+        final ITransmitter transmitter, //
+        final IReceiver receiver) //
+    {
+        super(peer, TransferType.INSTANT, projects, sessionManager, session,
+            editorManager, workspace, checksumCache, connectionService,
+            transmitter, receiver);
+    }
+
+    @Override
+    protected void setup(IProgressMonitor monitor) throws IOException,
+        LocalCancellationException {
+        if (fileTransferManager == null)
+            throw new LocalCancellationException(
+                "not connected to a XMPP server",
+                CancelOption.DO_NOT_NOTIFY_PEER);
+
+        /* get all opened editors */
+        editorManager.addSharedEditorListener(listener);
+        Set<SPath> openEditors = session.getComponent(
+            UserEditorStateManager.class).getOpenEditors();
+        for (SPath remoteOpenFile : openEditors)
+            fileOpened(remoteOpenFile);
+        for (SPath localOpenFile : editorManager.getOpenEditors())
+            fileOpened(localOpenFile);
+    }
+
+    @Override
+    protected void prepareTransfer(IProgressMonitor monitor,
+        List<FileList> fileLists) throws IOException,
+        SarosCancellationException {
+
+        /* until further patch, lock the complete session */
+        stoppedUsers = stopUsers(monitor);
+        sendAndAwaitActivityQueueingActivation(monitor);
+
+        remoteUser = session.getUser(getPeer());
+        if (remoteUser == null)
+            throw new LocalCancellationException(null,
+                CancelOption.DO_NOT_NOTIFY_PEER);
+
+        int fileCount = 0;
+        for (final FileList list : fileLists) {
+            fileCount += list.getPaths().size();
+
+            final String projectID = list.getProjectID();
+            final IProject project = session.getProject(projectID);
+
+            if (project == null)
+                throw new LocalCancellationException("project with id "
+                    + projectID + " was unshared during synchronization",
+                    CancelOption.NOTIFY_PEER);
+        }
+
+        createTransferList(fileLists, fileCount);
+        transmittedFiles = new HashSet<SPath>(fileCount * 2);
+    }
+
+    @Override
+    protected void transfer(IProgressMonitor monitor, List<FileList> fileLists)
```

Comment:
Are you aware that the current "INSTANT" mode does block the session even longer ?

1. I do not see any compression used at all.
2. We now have to wait until the remote side actually also receive all the data.

As already mentioned, the goal of this feature is to start collaborating as soon as possible, not on how fast the first file gets written to disk.

Reply:
Yes, but please read the commit message first. You find planed features also on the [Projectboard](https://github.com/saros-project/saros/projects/15).

This patch is the basement of further work.